node/src/types/appendable_block.rs

```diff
@@ -1,289 +1,243 @@
-                &self.transaction_config,
-                NO_LEEWAY,
-                self.timestamp,
-                transfer.deploy_hash(),
-            )
-            .is_err()
-        {
-            return Err(AddError::InvalidDeploy);
-        }
-        if self.has_max_transfer_count() {
-            return Err(AddError::TransferCount);
+        if footprint.ttl > self.transaction_config.max_ttl {
+            return Err(AddError::ExcessiveTtl);
         }
-        if self.would_exceed_approval_limits(transfer.approvals().len()) {
-            return Err(AddError::ApprovalCount);
-        }
-        self.deploy_and_transfer_set.insert(*transfer.deploy_hash());
-        self.total_approvals += transfer.approvals().len();
-        self.transfers.push(transfer);
-        Ok(())
-    }
-
-    /// Attempts to add a deploy to the block; returns an error if that would violate a validity
-    /// condition.
-    ///
-    /// This _must not_ be called with a transfer - the function cannot check whether the argument
-    /// is actually not a transfer.
-    pub(crate) fn add_deploy(
-        &mut self,
-        deploy: DeployHashWithApprovals,
-        footprint: &DeployFootprint,
-    ) -> Result<(), AddError> {
-        if self.deploy_and_transfer_set.contains(deploy.deploy_hash()) {
-            return Err(AddError::Duplicate);
+        if footprint.timestamp > self.timestamp {
+            return Err(AddError::FutureDatedDeploy);
         }
-        if footprint.header.expired(self.timestamp) {
+        let expires = footprint.timestamp.saturating_add(footprint.ttl);
+        if expires < self.timestamp {
             return Err(AddError::Expired);
         }
-        if footprint
-            .header
-            .is_valid(
-                &self.transaction_config,
-                NO_LEEWAY,
-                self.timestamp,
-                deploy.deploy_hash(),
-            )
-            .is_err()
+        let limit = match footprint.category {
+            TransactionCategory::Standard => self.transaction_config.block_max_standard_count,
+            TransactionCategory::Mint => self.transaction_config.block_max_mint_count,
+            TransactionCategory::Auction => self.transaction_config.block_max_auction_count,
+            TransactionCategory::InstallUpgrade => {
+                self.transaction_config.block_max_install_upgrade_count
+            }
+        };
+        // check total count by category
+        let category = footprint.category;
+        let count = self
+            .transactions
+            .iter()
+            .filter(|(_, item)| item.category == category)
+            .count();
+        if count.checked_add(1).ok_or(AddError::Count(category))? > limit as usize {
+            return Err(AddError::Count(category));
+        }
+        // check total gas
+        let gas_limit: U512 = self
+            .transactions
+            .values()
+            .map(|item| item.gas_limit.value())
+            .sum();
```

Comment:
This, together with `size` and `count` checks below, makes block validation quadratic in the number of deploys now. It might not be a huge performance hit, but we'll be doing millions of additions instead of thousands now, and I think we should be aware of it.

Reply:
Noted